### PR TITLE
feat: wildcard type restriction

### PR DESCRIPTION
**Pull request by adriantam**
## Description
Implement wildcard type restriction for DSL.

Screen Shot 2022-11-16 at 10 40 29 AM

## References
Close https://github.com/openfga/syntax-transformer/issues/93

## Review Checklist
- [X] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [X] The correct base branch is being used, if not `main`
- [X] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -27,7 +27,7 @@
   ],
   "author": "OpenFGA",
   "dependencies": {
-    "@openfga/sdk": "^0.1.0",
+    "@openfga/sdk": "0.1.1",
```
Comment (rhamzeh) [inline package.json:30]:
```suggestion
    "@openfga/sdk": "^0.1.1",
```

Do not pin dependencies on libraries unless you have to. This allows people to benefit from latest non-breaking changes even if you didn't update the dependencies.

File: src/check-dsl.ts [hunk 2]
```diff
@@ -62,7 +62,8 @@ const defaultError = (lines: string[]) => {
 // are tuple to user set.  If so, return the type and relationship.
 // Otherwise, return null as relationship
 function destructTupleToUserset(allowableType: string): string[] {
-  return allowableType.split("#", 2);
+  // For validation purpose, we don't care whether it is wildcard restricted or not
+  return allowableType.replace(":*", "").split("#", 2);
```
Comment (rhamzeh) [inline src/check-dsl.ts:66]: We do care though. If it allows `*`, it should not be allowed on a TupleToUserset, same as the case with a relation.

In both cases, we won't expand, so we should warn them

File: src/openfga.ne [hunk 3]
```diff
@@ -230,12 +236,16 @@ _no_relations   -> "none" (_newline):*
 _naming         -> (("$"):? ( [a-z] | [A-Z] | [0-9] |  "_" |  "-" ):+) {%
     data => data.flat(3).join('').trim()
 %}
+_allowed_naming         -> (("$"):? ( [a-z] | [A-Z] | [0-9] |  "_" |  "-" | "#" ):+ (":*"):?) {%
```
Comment (rhamzeh) [inline src/openfga.ne:239]: `$` should be only allowed on relations, not types. But let's have that conversation elsewhere. Not related to this PR.

File: tests/data/model-validation.ts [hunk 4]
```diff
@@ -779,15 +779,12 @@ type org
 `,
     expectedError: [
       {
-        endColumn: 23,
+        endColumn: 22,
```
Comment (rhamzeh) [inline tests/data/model-validation.ts:782]: These seem to have broken?

File: tests/data/model-validation.ts [hunk 5]
```diff
@@ -804,15 +801,12 @@ type org
...
     expectedError: [
       {
-        endColumn: 22,
+        endColumn: 21,
         endLineNumber: 6,
-        message: "Assignable relation 'member' must have types",
-        relatedInformation: {
-          type: "assignable-relation-must-have-type",
-        },
+        message: "Invalid syntax",
```
Comment (rhamzeh) [inline tests/data/model-validation.ts:806]: Same here

File: tests/data/model-validation.ts [hunk 6]
```diff
@@ -940,6 +934,27 @@ type group
...
+type user
+type group
+  relations
+    define member: [user, user:*:*]
+`,
+    expectedError: [
+      {
+        endColumn: 35,
+        endLineNumber: 6,
+        message: "Invalid syntax",
```
Comment (rhamzeh) [inline tests/data/model-validation.ts:950]: Shouldn't the message be more descriptive?

**Review by rhamzeh — CHANGES_REQUESTED** [review]

File: src/openfga.ne (hunk 3 above)
Comment (adriantam) [reply src/openfga.ne:239]: Let's track it via https://github.com/openfga/syntax-transformer/issues/96

File: tests/data/model-validation.ts (hunk 4 above)
Comment (adriantam) [reply tests/data/model-validation.ts:782]: Previously the grammar itself allows empty direct assignable type and we rely on the checkDSL script to see whether there are any types.  However, with the grammar update, the grammar itself now prevents user from specifying empty assignable types.  This makes the check cleaner.

File: tests/data/model-validation.ts (hunk 5 above)
Comment (adriantam) [reply tests/data/model-validation.ts:806]: See above - this is byproduct of grammar change which requires the types to be specified.

File: tests/data/model-validation.ts (hunk 6 above)
Comment (adriantam) [reply tests/data/model-validation.ts:950]: See above - this is byproduct of grammar change which requires the types to be specified.

File: package.json (hunk 1 above)
Comment (adriantam) [reply package.json]: updated

File: src/check-dsl.ts (hunk 2 above)
Comment (adriantam) [reply src/check-dsl.ts]: fixed

File: src/check-dsl.ts [hunk 7]
```diff
@@ -62,8 +62,9 @@ const defaultError = (lines: string[]) => {
 // are tuple to user set.  If so, return the type and relationship.
 // Otherwise, return null as relationship
 function destructTupleToUserset(allowableType: string): string[] {
-  // For validation purpose, we don't care whether it is wildcard restricted or not
-  return allowableType.replace(":*", "").split("#", 2);
+  const isWildcard = allowableType.includes(":*");
```
[the remark this answers is not in the document]
Comment (rhamzeh) [reply src/check-dsl.ts:65]: On the caller's end, the change would be from:

```
const [decodedType, decodedRelation, isWildcard] = destructTupleToUserset(item);
```

to

```
const { type: decodedType, relation: decodedRelation, isWildcard } = destructTupleToUserset(item);
```

[the remark this answers is not in the document]
Comment (adriantam) [reply src/check-dsl.ts:71]: updated

**Review by rhamzeh — APPROVED** [review]

**Pull request merged**